### PR TITLE
Add Fabric internal hooks

### LINKED ISSUE #17755: Fabric callback hooks and externally registered callbacks
### Description & Motivation

It is currently not possible to observe the lifecycle of a Fabric training loop unless the user adds callback logic to their code. At the minimum, it requires to import the callback and make a call to `fabric.call`:

```py
from my_package import MonitorCallback

fabric = Fabric(callbacks=[MonitorCallback()]

...
optimizer.step()
fabric.call("on_after_optimizer_step")
```

The need to modify the code is a limitation compared to the Trainer, which 1) can load [callbacks configured from entry points](https://lightning.ai/docs/pytorch/stable/extensions/entry_points.html#entry-points) and 2) has the training loop encapsulated, so the user doesn't need to call hooks.

### Pitch

1. Add the ability to load callbacks from entry points the same way Trainer does it.
2. Offer Fabric-specific callback methods for

- before and after `Fabric.setup()` gets called
- before and after optimizer step gets called
- before and after backward gets called

These hooks get called automatically by Fabric if the user defines the methods:

```py
if callable(getattr(callback, "on_after_optimizer_step", None)):
    callback.on_after_optimizer_step()
```

This will enable external applications to observe state and lifecycle of a Fabric training loop. 

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @tchaton @justusschock @awaelchli @carmocca

### PR DESCRIPTION
**Pull request by lightningforever**
## What does this PR do?

Fixes #17755

Adds Fabric-internal hooks that get called automatically.
This small selection of hooks remains undocumented/experimental for now. Subject to change in the future.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Looks good! We could add (not now is ok) an on_after_backward hook since it's the other event we control.

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
